Kzm s3 discovery report (#11)
Dasharo/twpm-docs: docs/tutorials/tutorial_start_EOS-S3_HW.md

and FPGA module.

### Development of the FPGA module

To develop the bitstream of the FPGA Synthesis the tool named _Symbiflow_
is used.

### Development of the ARM Cortex-M4

To develop the M4 part of the SoC standard ARM toolchain
could be used.

### Basic steps

To prepare development environment steps from [this instruction][build-instr-url]
should be used(parts _Docker image_, _Building_).

> Note: There was problem during compilation for Cortex-M4 device.

```shell
make -f makefiles/Makefile_Startup
make[2]: Entering directory '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project'
make[2]: *** No rule to make target '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project/output/startup_EOSS3B_GCC.o', needed by 'all'.  Stop.
make[2]: Leaving directory '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project'
make[1]: *** [Makefile:59: Startup] Error 2
make[1]: Leaving directory '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project'
make: *** [Makefile:54: m4] Error 2
```

There is workaround for this problem. During start of the Docker container

> one option should be corrected:

Reviewer: Please mention from which directory Docker should be started.

```shell
 -v $PWD:/home/qorc-sdk/workspace
```

to

```shell
 -v $PWD:/home/qorc-sdk/qorc-sdk/workspace
```

Warning: this correction reorganizes structure of the catalog in Docker container.

More generic solution is to pay attention about nesting the directories.

> Note: with newest environment, run by command:
```shell

> $ docker run --rm -it -v $PWD:/home/qorc-sdk/workspace \

Reviewer: ```suggestion
$ docker run --rm -it -v $PWD:/home/qorc-sdk/qorc-sdk/workspace \
```

    --device=/dev/ttyACM0:/dev/ttyS\_QORC ghcr.io/dasharo/twpm-sdk:main
```
After succesful download it stops during execution `make`, with error, there
is no `yosys` utility. There is workaround for this problem, by using older
version of the container.
```shell

> $ docker run --rm -it -v $PWD:/home/qorc-sdk/workspace \

Reviewer: Same as above.

>     --device=/dev/ttyACM0:/dev/ttyS\_QORC 8f6e8a2450a3080d1f6d121214b88bc461299c201c61205d373f3510b2dfc3b4

Reviewer: I'm getting error:

```
docker: Error response from daemon: No such image: sha256:8f6e8a2450a3080d1f6d121214b88bc461299c201c61205d373f3510b2dfc3b4.
```

This is probably an image you built locally, not a remote one.

Reviewer: `ghcr.io/dasharo/twpm-sdk@sha256:ca99beadbc692e921762fdbe478d79bdfad3afd3db8d3e4084041bd51caaf6af` works, as Maciej mentioned on Matrix.

```
With this workaround there is one problem solved: FPGA synthesis is succesful.
There is still problem with compiling code for M4 (probable problem

> with paths; `make` doesn't have proper rule for compiling C code).

Reviewer: Can you paste the exact error message? It builds for me, although there are some warnings along the way (incompatible pointer types, implicit function declarations).

Does `git status` show any changes to Makefiles perhaps?

Reviewer: I couldn't replicate the failure, so this paragraph was removed.

Entry point for this application is [main.c][snippet-main-entry].
Assigning the signal to pad has been done in the [qorc_hardwareSetup() function][snippet-hardware-setup-function].
To proper assignment the array `PadConfig pincfg_table[]`, passed as first
argument should be initialized [code here][snippet-fpga-pin-config].
Assigning pin to FPGA structure could be performed as in the following [example][snippet-pad-config].
For controlling GPIO via FPGA the example is [qorc_fpga_gpioctrl][snippet-gpio-control].

### FPGA-RAM Communication

This task has been depicted in the [qf_ramblockinit example][snippet-ramblockinit].

Communication between RAM and FPGA/M4 is performed by AHB bus. Initialization
is showed in the [following example][snippet-m4-ram-fpga-communication].

Access to already declared RAM has been instantiated in the following [example][snippet-access-to-ram-from-fpga].

### Examples: qorc_helloworldm4fpga, qorc_helloworldm4fpgaheader

These two examples are located here: [qorc_helloworldm4fpga][example-helloworldm4fpga]
and [qorc_helloworldm4fpgaheader][example-helloworldm4fpga_header].
These two examples are similar, in system behaviour (blinking red LED).

> The difference between those two projects is the dynamic loading bitstream

Reviewer: ```suggestion
The difference between those two projects is the dynamically loaded bitstream
```

> # Getting started with QuickLogic EOS S3 Board

Reviewer 2: Consider renaming file to e.g. `getting_started_eos_s3.md` so it better relates to the actual document

make[2]: Leaving directory '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project'
make[1]: *** [Makefile:59: Startup] Error 2
make[1]: Leaving directory '/home/qorc-sdk/workspace/qorc-onion-apps/qorc_fpga_breathectrl/GCC_Project'
make: *** [Makefile:54: m4] Error 2
```

There is workaround for this problem. During start of the Docker container
one option should be corrected:
```
 -v $PWD:/home/qorc-sdk/workspace
```
to 
```
 -v $PWD:/home/qorc-sdk/qorc-sdk/workspace 
```

Warning: this correction reorganizes structure of the catalog in Docker container.

More generic solution is to pay attention about nesting the directories. 

## Architecture overview

Architectural overview of this SoC has been depicted in document: 
[EOS S3 Datasheet](https://www.quicklogic.com/wp-content/uploads/2020/12/QL-EOS-S3-Ultra-Low-Power-multicore-MCU-Datasheet-2020.pdf), page 19.
 
There is also different block diagram of the S3, with main subsystems,
printed in [QL S3 Technical Reference Manual](https://www.quicklogic.com/wp-content/uploads/2020/06/QL-S3-Technical-Reference-Manual.pdf), page 23. 

> ## Example code

Reviewer: [qorc_fpga_timerctrl](https://github.com/coolbreeze413/qorc-onion-apps/tree/a960ca3f450fafde9da07547b35efe8e1caa2574/qorc_fpga_timerctrl) shows both data communication and interrupts.

> There is example applications, in :[Source](https://github.com/coolbreeze413/qorc-onion-apps/tree/master/). 

Reviewer: ```suggestion
There are example applications, in: [Source](https://github.com/coolbreeze413/qorc-onion-apps/tree/master/). 
```

Author: (corrected)

Bunch of them, with pattern: `qorc_helloworld*` are the examples of basic
snippets, which we could use as a template for further applications.

Currently, as example, I use `qorc_helloworldm4` app. This is
FreeRTOS-compatible, fully working snippet. 

This program uses the following modules:
* Basic ARM Cortex M4 initialization;
* Physical interfaces:
  * UART
  * I2C
* GPIO, RGB LED diode

### Pin mapping

Entry point for this application is [main.c](https://github.com/coolbreeze413/qorc-onion-apps/blob/master/qorc_helloworldm4/src/main.c).
Assigning the signal to pad has been done in the [qorc_hardwareSetup() function](https://github.com/coolbreeze413/qorc-onion-apps/blob/a960ca3f450fafde9da07547b35efe8e1caa2574/qorc_helloworldm4/src/qorc_hardwaresetup.c#LL48C9-L48C9).
To proper asignment the array `PadConfig pincfg_table[]`, passed as first
argument should be initialized [code here](https://github.com/coolbreeze413/qorc-onion-apps/blob/a960ca3f450fafde9da07547b35efe8e1caa2574/qorc_helloworldm4/src/pincfg_table.c#L21)

> Assigning pin to FPGA structure could be performed as in the following [example](https://github.com/coolbreeze413/qorc-onion-apps/blob/a960ca3f450fafde9da07547b35efe8e1caa2574/qorc_helloworldfpga/fpga/rtl/quickfeather.pcf#LL44C8-L44C8)

Reviewer: Is this enough? M4 part of SoC must be told to not drive those lines somehow, or not told to drive if that is the default state. This must be clearly stated here.
